components/org.wso2.carbon.identity.application.authenticator.facebook/src/main/java/org/wso2/carbon/identity/application/authenticator/facebook/FacebookExecutor.java

```diff
@@ -0,0 +1,338 @@
+            ClaimConfig claimConfig = context.getExternalIdPConfig().getIdentityProvider().getClaimConfig();
+            if (StringUtils.isBlank(claimConfig.getUserClaimURI())) {
+                claimConfig.setUserClaimURI(getEffectiveClaimUri(claimDialectUri,
+                        FacebookAuthenticatorConstants.EMAIL));
+            }
+
+            String userIdClaimURI = context.getExternalIdPConfig().getUserIdClaimUri();
+            if (StringUtils.isNotBlank(userIdClaimURI)) {
+                userIdClaimURI = getEffectiveClaimUri(claimDialectUri, userIdClaimURI);
+            }
+
+            Map<String, Object> mappedClaims = new HashMap<>();
+            for (Map.Entry<String, Object> entry : result.entrySet()) {
+                String claimUri = getEffectiveClaimUri(claimDialectUri, entry.getKey());
+                Object claimValueObject = entry.getValue();
+                if (StringUtils.isNotEmpty(claimUri) && claimValueObject != null && StringUtils.isNotEmpty(
+                        claimValueObject.toString())) {
+                    String mappedClaimUri = getMappedLocalClaim(claimConfig, claimUri);
+                    if (StringUtils.isNotBlank(mappedClaimUri)) {
+                        mappedClaims.put(mappedClaimUri, claimValueObject.toString());
+                    }
+                }
+                // Set user ID claim if available.
+                if (userIdClaimURI != null && userIdClaimURI.equals(entry.getKey())
+                        && StringUtils.isNotBlank(userIdClaimURI) && claimValueObject != null) {
+                    mappedClaims.put(USERNAME_CLAIM, claimValueObject.toString());
+                }
+            }
+
+            return mappedClaims;
+        } catch (IOException e) {
+            LOG.error("Failed to retrieve user info from Facebook.", e);
+            throw buildExecutorFailure("Failed to retrieve user info from Facebook", e);
+        }
+    }
+
+    protected String getMappedLocalClaim(ClaimConfig claimConfig, String claimUri) {
+
+        ClaimMapping mappedClaim = Arrays.stream(claimConfig.getClaimMappings()).filter(
+                        claimMapping -> claimMapping.getRemoteClaim().getClaimUri().equals(claimUri)).findFirst()
+                .orElse(null);
+        if (mappedClaim != null) {
+            return mappedClaim.getLocalClaim().getClaimUri();
+        }
+        return null;
+    }
+
+    private String getValidatedPortalUrl(FlowExecutionContext context) throws FlowEngineException {
+
+        String portalUrl = context.getPortalUrl();
+        if (StringUtils.isBlank(portalUrl)) {
+            LOG.error("Portal URL is missing in FlowExecutionContext.");
+            throw buildExecutorFailure("Portal URL is required but not provided.", null);
+        }
+        return portalUrl;
+    }
+
+    private FlowEngineServerException buildExecutorFailure(String message) {
+
+        return new FlowEngineServerException(
```

Comment:
since we are not using the error codes and message sent from this exception, shall we use the FlowEngineServerException with one argument and only send the error message?